matrix/matrix.py

```diff
@@ -0,0 +1,161 @@
+"""Matrix data type class."""
+
+import operator
+
+
+class Matrix:
+    """An immutable object representing a matrix data-type.
+
+    Attributes:
+        _matrix (tuple): contains the data about the matrix. Each line of
+            the matrix is represented by a tuple.
+    Note:
+        This implementation only supports squared matrices (n * n).
+    """
+
+    def __init__(self, lines):
+        self._validate(lines)
+        self._matrix = lines
+
+    def _validate(self, data: tuple):
+        """Check if the data given is valid.
+
+        Note:
+            Valid data is defined if it contains tuples of the same length
+            (Matrix only supports n * n), items can't be iterable.
+        """
+        if any(hasattr(item, '__len__') for line in data for item in line):
+            raise TypeError("Matrix item type not supported")
+
+        if not all([len(line) == len(data) for line in data]):
+            raise ValueError("Matrix must be even-sided")
+
+    @classmethod
+    def unity(cls, dimension: int):
+        """Generate a matrix with a main diagonal of 1's, elsewhere 0's."""
+        return cls(
+            tuple([tuple((0,) * i + (1,) + (0,) * (dimension - i - 1)) for i in
+                   range(dimension)]))
+
+    @classmethod
+    def ones(cls, dimension):
+        """Generate a matrix full of 1's."""
+        return cls(tuple(([tuple((1,) * dimension) for _ in
```

Comment:
`tuple((1,) * dimension)` is redundant, it's already a tuple
Also, you already have a shorter way to repeat objects (` * dimension`), why don't you use it?